src/mlpack/methods/random_forest/bootstrap.hpp

```diff
@@ -18,30 +18,234 @@ namespace mlpack {
+ */
+class IdentityBootstrap
+{
+ public:
+  template<bool UseWeights,
+           typename MatType,
+           typename LabelsType,
+           typename WeightsType>
+  void Bootstrap(const MatType& dataset,
+                 const LabelsType& labels,
+                 const WeightsType& weights,
+                 MatType& bootstrapDataset,
+                 LabelsType& bootstrapLabels,
+                 WeightsType& bootstrapWeights)
+  {
+    MakeAlias(bootstrapDataset, dataset, dataset.n_rows, dataset.n_cols);
+    MakeAlias(bootstrapLabels, labels, labels.n_elem);
+    if (UseWeights)
+      MakeAlias(bootstrapWeights, weights, weights.n_elem);
+  }
+};
+
+/**
+ * Sequential bootstrap.
+ * 
+ * This sequential bootstrap is suitable for time-series that are not
+ * independent and identically distributed (IID). The bootstrap will
+ * generate variations of the data that are more IID. In order to do
+ * that it takes an indicator matrix that is a measure for how
+ * much informational overlap exists between each data point in the
+ * data set. Then it will randomly draw from the dataset with
+ * replacement but reducing the likelihood to draw data points that
+ * have high informational overlap with already drawn samples.
+ *
+ * @tparam IndMatType Indicator matrix type.
+ */
+template<typename IndMatType = arma::umat>
+class SequentialBootstrap
+{
+ public:
+  /**
+   * Constructor.
+   *
+   * @param[in] intervals See ComputeAverageUniqueness().
+   */
+  SequentialBootstrap(const IndMatType& intervals) :
+    intervals(intervals)
+  {
+    if (intervals.n_rows != 2)
+      throw std::invalid_argument(
+        "SequentialBootstrap::SequentialBootstrap(): "
+        "intervals must be a 2 x m matrix!");
+  }
+
+  /**
+   * Compute the average uniqueness of each event at any sampling time point.
+   * The average uniqueness is a measure for how isolated an event is during
+   * its lifetime from other events.
+   *
+   * @param[in] intervals Is a `2 x m` matrix, where each of the m columns has
+   *                      the start sample and the end sample of an interval.
+   *                      This matrix is a space-efficient form of the
+   *                      indicator matrix.
+   * @param[in] indices Indices of the average uniqueness that is returned.
+   *                    This parameter is used for optimization purposes so
+   *                    that in the unit tests the average uniqueness for
+   *                    all events can be accessed while in the production
+   *                    code only a single element is accessed.
+   * @return The average uniqueness of the events in @p intervals.
+   */
+  static arma::vec ComputeAverageUniqueness(
+    const IndMatType& intervals,
+    const arma::uvec& indices,
+    const arma::vec&  concurrency)
+  {
+    arma::vec       avg(indices.n_rows, arma::fill::zeros);
+    const arma::vec invConcurrency(
+      arma::vec(concurrency.n_rows, arma::fill::ones) / concurrency);
```

Comment:
I guess that in this function, the concurrency is dependent on the previous samples (phi), as it is computed only on the samples we have already chosen, plus the "candidate" sample (the last element of phi).  If we want to compute the concurrency in constant time, we can simply take the concurrency of all elements in phi without the "candidate" and add 1 to each interval that contains the 'candidate' point.  For this we would need to loop over `intervals`.

But it doesn't look like that's what is happening here---did I misunderstand the strategy or is there an optimization that I overlooked?

Reply:
Oh!  Actually.  Maybe there is a small issue in the Python code you wrote.

```
def getAvgUniqueness(indM):
  # average uniqueness from indicator matrix
  c = indM.sum(axis=1) # concurrency
  u = indM.div(c, axis=0) # uniqueness
  avgU = u[u>0].mean() # average uniqueness
  return avgU
```

In this code, if `indM` has shape `(n_intervals, n_samples)`, then should the sum be over `axis=0` and the division over `axis=1`?  That seems to be what the test code you wrote indicates, but I want to double check.

Reply:
Yes, it is over `axis=0` _and_ `axis=1`